Generic SideEffectStorage

mapokko opened the pull request:
Summary

 Allows the implementing of the `SideEffectStorage` trait from a generic concrete type. The changes ensure that side effect storage.

  Key Changes

  1. Made Side Effect Trait Generic (commit ef45f72)

  - Updated SideEffectStorage trait to be generic over the side effect type T
  - Added generic type parameter T to PostgresRepository and PostgresTransaction
  - Modified trait bounds to properly constrain side effect types with SideEffect<SideEffectId = Uuid> + 
  Pickle + Send + Sync
  - Updated all repository and transaction method signatures to use the generic aggregate type

  2. Completed Generic Implementation (commit c8f50bd)

  - Fixed outbox implementation to work with the new generic constraints
  - Updated TransactionOutboxExt to work with T::SideEffect instead of raw side effect types
  - Modified RepositoryOutboxExt trait to properly handle aggregate-specific side effects
  - Updated all test helpers and examples to specify the aggregate type parameter
  - Added proper trait bounds ensuring aggregate types implement required traits

  3. Code Formatting (commit 7009ced)

  - Fixed formatting issue in transaction.rs where clause

aewakefield committed:
wip: side effect trait must be generic

mapokko committed:
fix: finish updating generic

mapokko committed:
fix: format

aewakefield commented on eventastic_outbox_postgres/src/outbox.rs:
```diff
@@ -101,15 +106,22 @@ where
+    T: Aggregate<AggregateId = Uuid> + Send + Sync + Pickle + 'static,
+    T::SideEffect: SideEffect + Pickle + Clone + Send + Sync + 'static,
+    <T::SideEffect as SideEffect>::SideEffectId: Clone + Send + 'static,
+    T::SideEffect: SideEffect<SideEffectId = Uuid> + Pickle + Send + Sync,
```
Think these are duplicates.

```suggestion
    T::SideEffect: SideEffect<SideEffectId = Uuid> + Pickle + Send + Sync,
```

aewakefield commented on eventastic_postgres/src/transaction.rs:
```diff
@@ -21,21 +21,22 @@ use sqlx::{Postgres, Transaction};
     pub(crate) outbox: &'a O,
     pub(crate) tables: &'a TableRegistry,
     pub(crate) encryption_provider: &'a E,
+    pub(crate) phantom_side_effect: std::marker::PhantomData<T>,
```
```suggestion
    pub(crate) phantom_aggregate: std::marker::PhantomData<T>,
```

aewakefield commented on eventastic_postgres/src/repository.rs:
```diff
@@ -21,20 +23,24 @@ use sqlx::{
     pub(crate) outbox: O,
     pub(crate) tables: TableRegistry,
     encryption_provider: E,
+    phantom_side_effect: std::marker::PhantomData<T>,
```
```suggestion
    phantom_aggregate: std::marker::PhantomData<T>,
```

aewakefield committed:
chore: tidy up

jdon committed:
Remove table registry

Co-Authored-By: Alex Wakefield <1181370+aewakefield@users.noreply.github.com>
Co-Authored-By: Fabio Mirza <79943909+mapokko@users.noreply.github.com>

aewakefield committed:
chore: simplify the error message

When we get this error logged we end up with lots of different
messages due to the aggregate ID and version number. These should be
logged outside of the message to make grouping this easier.

jdon committed:
Merge pull request #17 from jdon/simplify_error_message

chore: simplify the error message

jdon committed:
Merge pull request #16 from jdon/side-effects-changes

Side effects changes
